R/tfb-wavelet-utils.R

```diff
@@ -0,0 +1,140 @@
+remove_slope <- function(x, y) {
+  last_element <- length(x)
+  slope <- (y[last_element] - y[1]) / (x[last_element] - x[1])
+  intercept <- y[1] - slope * x[1]
+  f <- intercept + slope * x
+  y_desloped <- y - f
+  y_desloped <- structure(f,
+                          intercept = intercept,
+                          slope = slope)
+  y_desloped
+}
+
+
+interpolate_arg <- function(arg_list) {
+  n <- vapply(arg_list, length, numeric(1))
+  closest_power <- round(log2(n))
+  n_diff <- 2^closest_power - n
+  
+  interp_index <- map2(arg_list, 
+                       n_diff,
+                       function(x, y) {
+                         slope <- (max(x) - min(x)) / (length(x) + y - 1)
+                         seq(min(x), max(x), by = slope)
+                       })
+  unlist(interp_index)
+}
+
+
+
+fit_wavelet <- function(data, Z, penalized, glmnet_args) {
+  eval_list <- split(data$data, data$id)
+  arg_list <- split(data$arg, data$id)
+  
+  correction <- map2(arg_list, eval_list, function(x, y) remove_slope(x,y))
+  
+  if (penalized) {
+    coefs <- map2(eval_list, correction, 
+                  function(x, y) {
+                    Z <- cbind(1, Z)
+                    # Least squares fit, directly computed
+                    as.vector((t(Z) %*% (x - y)) / diag(crossprod(Z)))})
+  } else {
+    coefs <- map2(eval_list, correction, 
+                  function(x, y) {
+                    temp_model <- do.call(glmnet::cv.glmnet, list(Z, x - y, 
+                                                                  glmnet_args))
```

Comment:
this won't work:

```r
tfb_wavelet(woo_tfd, penalized = TRUE, level = 8)
Error in glmnet(x, y, weights = weights, offset = offset, lambda = lambda,  : 
  number of elements in weights (1) not equal to the number of rows of x (256)
```
because you get a list in which glmnet_args is a sublist ---
do `c(list(x = Z, y = x), glmnet_args)` instead.